### PR TITLE
Preliminary support for `HTTP_AUTH`.

### PR DESCRIPTION
**Pull request by mattmoor**
I am still working through a few issues, but I believe the main problem at this point is that we don't support range requests.

### CODE REVIEW

File: apko/private/apk.bzl [hunk 1]
```diff
@@ -15,6 +15,29 @@ filegroup(
 )
 """
 
+def _auth(rctx, url):
+    http_auth = rctx.getenv("HTTP_AUTH")
```
Comment (mattmoor) [inline apko/private/apk.bzl:19]: It looks like this only works in more recent versions of Bazel.

@thesayyn @alexeagle what's the best way of doing this in a portable way?

**Comment by mattmoor** [comment]
It looks like Alpine withdrew a package breaking one of the legs?
```
Error in download: java.io.IOException: Error downloading [https://dl-cdn.alpinelinux.org/alpine/v3.18/main/aarch64/libcrypto3-3.1.5-r0.apk] to /home/runner/.bazel/external/_main~apko~examples_multi_arch_and_repo_libcrypto3_aarch64_3.1.5-r0/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fv3.18%2Fmain/aarch64/libcrypto3-3.1.5-r0/06951e2d641fbcc40718707542c45a30ca1807f6.sig.tar.gz: GET returned 404 Not Found
```

**Comment by imjasonh** [comment]
> It looks like Alpine withdrew a package breaking one of the legs?
> 
> ```
> Error in download: java.io.IOException: Error downloading [https://dl-cdn.alpinelinux.org/alpine/v3.18/main/aarch64/libcrypto3-3.1.5-r0.apk] to /home/runner/.bazel/external/_main~apko~examples_multi_arch_and_repo_libcrypto3_aarch64_3.1.5-r0/https%3A%2F%2Fdl-cdn.alpinelinux.org%2Falpine%2Fv3.18%2Fmain/aarch64/libcrypto3-3.1.5-r0/06951e2d641fbcc40718707542c45a30ca1807f6.sig.tar.gz: GET returned 404 Not Found
> ```

We should probably migrate tests to only pull from Wolfi, where we withdraw less often and have more control in general.

**Comment by mattmoor** [comment]
This fixes alpine for now by just regenerating the lock, but I agree we should switch to something where we have more control: https://github.com/chainguard-dev/rules_apko/pull/85

File: apko/private/apk.bzl (hunk 1 above)
Comment (alexeagle) [reply apko/private/apk.bzl:19]: There is history on rules_oci for this: we carried a fallback of spawning 'curl' for this

**Review by imjasonh — APPROVED** [review]

**Comment by sfc-gh-mhazy** [comment]
@mattmoor Thank you for the changes. We are quite excited about using this instead of building on top of base image!

I am wondering if we could use similar approach for fetching token within bazel rules as in [rules_oci](https://github.com/bazel-contrib/rules_oci/blob/8042fcbb664ac06ae129b027891b1024ac4e10dd/oci/private/authn.bzl#L277) so that the rule don't depend on external env vars. I don't know much details about chainctl, but I am happy to implement the auth in bazel if you think this is feasible.

**Comment by mattmoor** [comment]
@sfc-gh-mhazy I'm receptive to that.  I added `HTTP_AUTH` mostly for symmetry with `apko`, `melange` and `apk`, but we'd love something better.

I bet @imjasonh and @jonjohnsonjr will have thoughts, but open to ideas for what we'd do here!

**Comment by sfc-gh-mhazy** [comment]
@imjasonh I think ideally the bazel rules would
1. Not depend on chainctl being on host 
   * If feasible, fetch the token as in rules_oci, without chainctl command at all. rules_oci seem to work for private chainguard images registries, not sure if this is a good sign here. 
   * we could add chainctl toolchain (similar as apko toolchain)
2. Don't fetch the token as part of build, because we want to support "offline" mode. 
3. We will need to think how to work with `apko_lock` which is not hermetic and serves for refreshing purposes. 

In rules_apko, the apk fetch happens completely outside of apko, as part of [apk.bzl](https://github.com/chainguard-dev/rules_apko/blob/main/apko/private/apk.bzl). Perhaps we should fetch token there ("natively" with rctx.download or with chainctl)

Maybe lets move the discussion to some doc?

**Comment by imjasonh** [comment]
Thanks, this is really great context.

I think the fix will be for apk.bzl to fetch the token (using chainctl or `CHAINGUARD_IDENTITY` env var, either way), and populate the cache for offline image builds.

Then, when building the image in offline mode, we should (we might already) not attempt to get a token until we're actually about to use it to fetch an apk -- which we won't do if offline mode. I think this might already work this way, or if not, it could fairly easily. This would be a performance benefit even in online-mode, since a token wouldn't be requested if there are no APKs to fetch if they're already in the cache.

**Pull request merged**